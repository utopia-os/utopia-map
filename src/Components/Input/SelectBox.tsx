import InformationCircleIcon from '@heroicons/react/24/outline/InformationCircleIcon'
import { useState } from 'react'

interface SelectBoxProps {
  labelTitle?: string
  labelStyle?: string
  type?: string
  containerStyle?: string
  defaultValue: string
  placeholder?: string

  updateFormValue: (value: string) => void

  options: { name: string; value: string }[]
  labelDescription?: string
}

/**
 * @category Input
 */
export function SelectBox(props: SelectBoxProps) {
  const {
    labelTitle,
    labelDescription,
    defaultValue,
    containerStyle,
    placeholder,
    labelStyle,
    options,
    updateFormValue,
  } = props

  const [value, setValue] = useState(defaultValue || '')

  const updateValue = (newValue: string) => {
    updateFormValue(newValue)
    setValue(newValue)
  }

  return (
    <div className={`tw-inline-block ${containerStyle ?? ''}`}>
      {labelTitle ? (
        <label className={`tw-label  ${labelStyle ?? ''}`}>
          <div className='tw-label-text'>
            {labelTitle}
            {labelDescription && (
              <div className='tw-tooltip tw-tooltip-right' data-tip={labelDescription}>
                <InformationCircleIcon className='tw-w-4 tw-h-4' />
              </div>
            )}
          </div>
        </label>
      ) : (
        ''
      )}
      <select
        className='tw-select tw-select-bordered tw-w-full'
        value={value}
        onChange={(e) => updateValue(e.target.value)}
      >
        <option disabled value='PLACEHOLDER'>
          {placeholder}
        </option>
        {options.map((o, k) => {
          return (
            <option value={o.value || o.name} key={k}>
              {o.name}
            </option>
          )
        })}
      </select>
    </div>
  )
}
