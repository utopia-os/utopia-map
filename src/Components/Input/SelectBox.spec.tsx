import { render, screen, fireEvent } from '@testing-library/react'
import { describe, it, expect, beforeEach, vi } from 'vitest'

import { SelectBox } from './SelectBox'

describe('<SelectBox />', () => {
  let wrapper: ReturnType<typeof render>

  const updateFormValue = vi.fn()

  beforeEach(() => {
    vi.clearAllMocks()
    wrapper = render(
      <SelectBox
        defaultValue={'option-1'}
        updateFormValue={updateFormValue}
        options={[
          {
            name: 'Option 1',
            value: 'option-1',
          },
          {
            name: 'Option 2',
            value: '',
          },
        ]}
      />,
    )
  })

  it('renders properly', () => {
    expect(wrapper.container.firstChild).toMatchSnapshot()
  })

  describe('handleChange', () => {
    it('calls updateFormValue with new value', () => {
      fireEvent.change(screen.getByRole('combobox'), { target: { value: 'option-1' } })
      expect(updateFormValue).toBeCalledWith('option-1')
    })
  })

  describe('labelTitle', () => {
    it('sets label', () => {
      wrapper.rerender(
        <SelectBox
          defaultValue={''}
          updateFormValue={updateFormValue}
          labelTitle='My Title'
          options={[]}
        />,
      )
      expect(wrapper.container.firstChild).toMatchSnapshot()
    })
  })

  describe('labelDescription', () => {
    it('sets label description', () => {
      wrapper.rerender(
        <SelectBox
          defaultValue={''}
          updateFormValue={updateFormValue}
          labelTitle='My Title'
          labelDescription='My Description'
          options={[]}
        />,
      )
      expect(wrapper.container.firstChild).toMatchSnapshot()
    })
  })
})
