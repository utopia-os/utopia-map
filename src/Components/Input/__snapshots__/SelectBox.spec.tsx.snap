// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SelectBox /> > labelDescription > sets label description 1`] = `
<div
  class="tw-inline-block "
>
  <label
    class="tw-label  "
  >
    <div
      class="tw-label-text"
    >
      My Title
      <div
        class="tw-tooltip tw-tooltip-right"
        data-tip="My Description"
      >
        <svg
          aria-hidden="true"
          class="tw-w-4 tw-h-4"
          data-slot="icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m11.25 11.25.041-.02a.75.75 0 0 1 1.063.852l-.708 2.836a.75.75 0 0 0 1.063.853l.041-.021M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0Zm-9-3.75h.008v.008H12V8.25Z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </div>
    </div>
  </label>
  <select
    class="tw-select tw-select-bordered tw-w-full"
  >
    <option
      disabled=""
      value="PLACEHOLDER"
    />
  </select>
</div>
`;

exports[`<SelectBox /> > labelTitle > sets label 1`] = `
<div
  class="tw-inline-block "
>
  <label
    class="tw-label  "
  >
    <div
      class="tw-label-text"
    >
      My Title
    </div>
  </label>
  <select
    class="tw-select tw-select-bordered tw-w-full"
  >
    <option
      disabled=""
      value="PLACEHOLDER"
    />
  </select>
</div>
`;

exports[`<SelectBox /> > renders properly 1`] = `
<div
  class="tw-inline-block "
>
  <select
    class="tw-select tw-select-bordered tw-w-full"
  >
    <option
      disabled=""
      value="PLACEHOLDER"
    />
    <option
      value="option-1"
    >
      Option 1
    </option>
    <option
      value="Option 2"
    >
      Option 2
    </option>
  </select>
</div>
`;
